File: zenoh-java/src/commonMain/kotlin/io/zenoh/qos/Priority.kt
//
// Copyright (c) 2023 ZettaScale Technology
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
//
// Contributors:
//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
//

package io.zenoh.qos

/**
 * The Priority of Zenoh messages.
 *
 * A Priority is identified by a numeric value. Lower the value, higher the priority. Higher the value, lower the priority.
 *
 * - Highest priority: 1 ([REALTIME])
 * - Lowest priority: 7 ([BACKGROUND])
 */
enum class Priority(internal val jni: io.zenoh.jni.qos.Priority) {
    REALTIME(io.zenoh.jni.qos.Priority.REAL_TIME),
    INTERACTIVE_HIGH(io.zenoh.jni.qos.Priority.INTERACTIVE_HIGH),
    INTERACTIVE_LOW(io.zenoh.jni.qos.Priority.INTERACTIVE_LOW),
    DATA_HIGH(io.zenoh.jni.qos.Priority.DATA_HIGH),
    DATA(io.zenoh.jni.qos.Priority.DATA),
    DATA_LOW(io.zenoh.jni.qos.Priority.DATA_LOW),
    BACKGROUND(io.zenoh.jni.qos.Priority.BACKGROUND);

    val value: Int
        get() = jni.value

    companion object {
        fun fromInt(value: Int) = entries.first { it.value == value }

        internal fun fromJni(jni: io.zenoh.jni.qos.Priority): Priority = fromInt(jni.value)
    }
}

File: zenoh-flat/src/qos/priority.rs
use prebindgen_proc_macro::prebindgen;

/// Message priority policy. Lower numeric value means higher priority.
#[prebindgen]
#[repr(i32)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Priority {
    RealTime = 1,
    InteractiveHigh = 2,
    InteractiveLow = 3,
    DataHigh = 4,
    Data = 5,
    DataLow = 6,
    Background = 7,
}

impl From<zenoh::qos::Priority> for Priority {
    fn from(value: zenoh::qos::Priority) -> Self {
        match value {
            zenoh::qos::Priority::RealTime => Priority::RealTime,
            zenoh::qos::Priority::InteractiveHigh => Priority::InteractiveHigh,
            zenoh::qos::Priority::InteractiveLow => Priority::InteractiveLow,
            zenoh::qos::Priority::DataHigh => Priority::DataHigh,
            zenoh::qos::Priority::Data => Priority::Data,
            zenoh::qos::Priority::DataLow => Priority::DataLow,
            zenoh::qos::Priority::Background => Priority::Background,
        }
    }
}

impl From<Priority> for zenoh::qos::Priority {
    fn from(value: Priority) -> Self {
        match value {
            Priority::RealTime => zenoh::qos::Priority::RealTime,
            Priority::InteractiveHigh => zenoh::qos::Priority::InteractiveHigh,
            Priority::InteractiveLow => zenoh::qos::Priority::InteractiveLow,
            Priority::DataHigh => zenoh::qos::Priority::DataHigh,
            Priority::Data => zenoh::qos::Priority::Data,
            Priority::DataLow => zenoh::qos::Priority::DataLow,
            Priority::Background => zenoh::qos::Priority::Background,
        }
    }
}
